### PR TITLE
[5.4] Actually skip not needed dependencies during dependency resolution

### PR DESCRIPTION
**Pull request by neonichu**
5.4 cherry-pick of #3166

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please test

**Review by abertelrud — APPROVED** [review]

**Comment by tomerd** [comment]
cc @airspeedswift

**Review by tomerd — APPROVED** [review]

File: Sources/PackageModel/Manifest.swift [hunk 1]
```diff
@@ -165,21 +165,21 @@ public final class Manifest: ObjectIdentifierProtocol {
...
             for targetDependency in target.dependencies {
                 if let dependency = packageDependency(referencedBy: targetDependency) {
-                    requiredDependencies.insert(dependency)
+                    requiredDependencyURLs.insert(dependency.url)
                 }
             }
         }
-
-        return Array(requiredDependencies)
+
```
Comment (federicobucchi) [inline Sources/PackageModel/Manifest.swift:181]: Can we please remove the added invisibles at line 171, 173 and 181?

**Review by federicobucchi — APPROVED** [review]

File: Sources/PackageModel/Manifest.swift (hunk 1 above)
Comment (neonichu) [reply Sources/PackageModel/Manifest.swift:181]: We can surely remove them in main, but I don't think another multi-hour CI run is necessary for this to do it in 5.4 as well.

**Review by abertelrud — APPROVED** [review]

**Comment by neonichu** [comment]
Scope of Issue: Any package that relied on dependencies not used by any products being skipped, e.g. because it is using a non-versioned dependency for tests. Notably, this is utilized by the Firebase SDK which is widely used.
Risk: Medium. This is changing dependency resolution, an area for which SwiftPM's test coverage isn't comprehensive. However, this is almost verbatim reverting the behavior to the one from 5.3, so it's unlikely to introduce any new issues.
Testing: I resurrected `testTargetBasedDependency` for this which was being skipped and also tested manually with a few projects. Specifically https://github.com/firebase/firebase-ios-sdk was broken before because it relies on test-only dependencies and works now.

**Review by airspeedswift — APPROVED** [review]

**Pull request merged**